proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -2673,36 +2817,29 @@ lemma cancel_all_invs'_helper:
                   else setThreadState Structures_H.thread_state.Inactive t
                od) q
    \<lbrace>\<lambda>rv. all_invs_but_ct_not_inQ'\<rbrace>"
+  supply if_split[split del]
   apply (rule mapM_x_inv_wp2)
    apply clarsimp
   apply (rule hoare_pre)
-   apply (wp valid_irq_node_lift valid_irq_handlers_lift'' irqs_masked_lift
-             hoare_vcg_const_Ball_lift untyped_ranges_zero_lift
-             sts_valid_queues sts_st_tcb' setThreadState_not_st
-        | simp add: cteCaps_of_def o_def)+
-(*   apply (unfold fun_upd_apply Invariants_H.tcb_st_refs_of'_simps)
-  apply clarsimp
+   apply (wpsimp wp: valid_irq_node_lift valid_irq_handlers_lift'' irqs_masked_lift hoare_vcg_const_Ball_lift
+                     sts_valid_queues sts_st_tcb' setThreadState_not_st possibleSwitchTo_sch_act_not_other)
+       apply (strengthen weak_sch_act_wf_D1)
+       apply (wpsimp wp: valid_irq_node_lift hoare_vcg_const_Ball_lift
+                         sts_valid_queues sts_st_tcb' setThreadState_not_st sts_sch_act'
+                  split: if_splits
+               simp_del: comp_apply)+
+     apply (wp hoare_drop_imp)
+    apply (wpsimp wp: hoare_vcg_const_Ball_lift hoare_vcg_all_lift gts_wp' hoare_vcg_imp_lift
+                      replyUnlink_valid_objs' replyUnlink_st_tcb_at'
+                simp: valid_tcb_state'_def simp_del: comp_apply)+
+  apply (clarsimp simp: sch_act_wf_weak_sch_act_wf valid_pspace_valid_objs' valid_objs'_valid_tcbs'
+                  cong: conj_cong)
   apply (intro conjI)
-  apply (clarsimp simp: valid_tcb_state'_def global'_no_ex_cap
-                 elim!: rsubst[where P=sym_refs]
-                 dest!: set_mono_suffix
-                intro!: ext
-       | (drule (1) bspec, clarsimp simp: valid_pspace'_def valid_tcb'_def elim!: valid_objs_valid_tcbE'))+ *)
-  sorry
-
-lemma ep_q_refs_max:
-  "\<lbrakk> ko_at' r p s; sym_refs (state_refs_of' s); r \<noteq> IdleEP \<rbrakk>
-      \<Longrightarrow> (state_refs_of' s p \<subseteq> (set (epQueue r) \<times> {EPSend, EPRecv}))
-       \<and> (\<forall>x\<in>set (epQueue r). \<exists>ntfnptr. state_refs_of' s x \<subseteq>
-                                  {(p, TCBBlockedSend), (p, TCBBlockedRecv), (ntfnptr, TCBBound)})"
-  apply (frule(1) sym_refs_ko_atD')
-  apply (drule ko_at_state_refs_ofD')
-  apply (case_tac r)
-  sorry (*
-    apply (clarsimp simp: st_tcb_at_refs_of_rev' tcb_bound_refs'_def
-             | rule conjI | drule(1) bspec | drule st_tcb_at_state_refs_ofD'
-             | case_tac ntfnptr)+
-  done *)
+    apply (clarsimp simp: global'_no_ex_cap)
+   apply clarsimp
+  apply (erule valid_queues_not_runnable'_not_ksQ)
+  apply (fastforce simp: pred_tcb_at'_def obj_at'_def projectKOs)
+  done
```

Comment:
Very nice proof! Having seen what this looked like halfway through the process, this was some great cleanup.